### ISSUE TITLE
substr_replace works incorrectly with utf-8

### ISSUE THREAD

**Issue by danyaPostfactum** [futtta/autoptimize#154]
`substr_replace` inserts string at wrong place under certain conditions. I found that autoptimize breaks my html. It looks like `substr_replace` works bad with UTF-8. You can see on the screenshot that `<script>` is inserted at wrong place. 
![1525757520075](https://user-images.githubusercontent.com/1527097/39766522-78338860-531f-11e8-9128-2e8eff048715.png)

I fixed this bug by replacing builtin `substr_replace` with custom `mb_substr_replace`:
```php
function mb_substr_replace($original, $replacement, $position, $length)
{
 $startString = mb_substr($original, 0, $position, "UTF-8");
 $endString = mb_substr($original, $position + $length, mb_strlen($original), "UTF-8");

 $out = $startString . $replacement . $endString;

 return $out;
}
````

I can provide my html before substr_replace for testing if you need. But I guess this bug depends on server configuration.

**Comment by zytzagoo** [comment]
Just cross-referencing this, since it's related: https://wordpress.org/support/topic/weird-bug-with-autoptimize-script-insertion/

If #155 is too much, perhaps we should add an entry to the FAQ related to this, mentioning the php.ini setting/change that can solve this without all the extra code:

```
mbstring.func_overload = 2
mbstring.internal_encoding = UTF-8
```

**Comment by futtta** [comment]
Are you sure the reporter in https://wordpress.org/support/topic/weird-bug-with-autoptimize-script-insertion/ is **adding** that to his php.ini? Or is he removing it?

If he is removing it; super and then indeed we should go the FAQ way. If he added it, then we'll _have_ to add your new code, because [mbstring.func_overload is deprecated and will get removed at some point](http://php.net/manual/en/mbstring.overload.php) .. :-/

**Comment by zytzagoo** [comment]
must be adding, beacuse that way strpos starts working properly for multibyte strings...

without overload, strpos gets wrong position if there are multibyte chars present preceding the searched string...

**Comment by futtta** [comment]
let try pinging him here to confirm; @Enuriru are you the same person who reported https://wordpress.org/support/topic/weird-bug-with-autoptimize-script-insertion/? And if so; can you confirm you **added** those 2 lines to your php.ini?

(and if not; sorry for bothering you :-) )

**Comment by futtta** [comment]
thanks for the feedback @Enuriru !

guess we have no choice but to bite the bullet @zytzagoo :-)

**Comment by futtta** [comment]
OK, so @Enuriru and @danyaPostfactum : 

As `mbstring.func_overload = 2` is deprecated we cannot rely on this for future use (guess as of PHP 7.3), so the code changes by @zytzagoo should make AO work for you even with that function and is more robust then the change @danyaPostfactum made. Could you [download the beta zipfile](https://github.com/futtta/autoptimize/archive/beta.zip) and test to confirm (or not)?

**Comment by futtta** [comment]
any feedback @Enuriru or @danyaPostfactum ? :-)

**Comment by danyaPostfactum** [comment]
Unfortunately I can't check, I have no longer access to the website..

**Comment by futtta** [comment]
no problem @danyaPostfactum , thanks for the feedback anyway! :+1: 

and if you have some time, feel free to take Autoptimize 2.4 beta for a spin elsewhere! :-)

**Comment by futtta** [comment]
Closing as fixed with AO24 release, although for 2.4.1 this will be off by default (but can be turned on with a filter provided mb_* functions are available).
